test/tools/d_do_test.d

```diff
@@ -962,11 +970,145 @@ int runBashTest(string input_dir, string test_name)
+    string output_dir, string output_file)
+{
+    const testScriptDir = buildPath(dmdTestDir, input_dir);
+    const testScriptPath = buildPath(testScriptDir, test_name ~ ".d");
+    const testOutDir = buildPath(output_dir, test_name);
+
+    writefln(" ... %s/%s.d", input_dir, test_name);
+
+    removeIfExists(output_file);
+    if (exists(testOutDir))
+        rmdirRecurse(testOutDir);
+    mkdirRecurse(testOutDir);
+
+    // create the "dshell" module for the tests
+    {
+        auto dshellFile = File(buildPath(testOutDir, "dshell.d"), "w");
+        dshellFile.writeln(`module dshell;
+public import dshell_prebuilt;
+static this()
+{
+    dshellPrebuiltInit("` ~ input_dir ~ `", "`, test_name , `");
+}
+`);
+    }
+
+    const testScriptExe = buildPath(testOutDir, "run" ~ envData.exe);
+    const output_file_temp = output_file ~ ".tmp";
+    mkdirsFor(output_file_temp);
+
+    int exitCode;
+    {
+        auto outfile = File(output_file_temp, "w");
+        const compile = [envData.dmd, "-conf=", "-m"~envData.model] ~
+            getPicFlags ~ [
+            "-od" ~ testOutDir,
+            "-of" ~ testScriptExe,
+            "-I=" ~ testScriptDir,
+            "-I=" ~ testOutDir,
+            "-I=" ~ buildPath(dmdTestDir, "tools", "dshell_prebuilt"),
+            "-i",
+            // Causing linker errors for some reason?
+            "-i=-dshell_prebuilt", getDShellPrebuilt(envData),
+            testScriptPath,
+        ];
+        outfile.writeln("[COMPILE_TEST] ", escapeShellCommand(compile));
+        // Note that spawnprocess closes the file, so it will need to be re-opened
+        // below when we run the test
+        auto compileProc = std.process.spawnProcess(compile, stdin, outfile, outfile);
+        exitCode = wait(compileProc);
+    }
+    if (exitCode == 0)
+    {
+        auto outfile = File(output_file_temp, "a");
+        const runTest = [testScriptExe];
+        outfile.writeln("[RUN_TEST] ", escapeShellCommand(runTest));
+        auto runTestProc = std.process.spawnProcess(runTest, stdin, outfile, outfile);
+        exitCode = wait(runTestProc);
+    }
+
+    if (exitCode != 0)
+    {
+        writeln("==============================");
+        writefln("Test '%s/%s.d' failed. The logged output:", input_dir, test_name);
+        const output = readText(output_file_temp);
+        write(output);
+        if (!output.endsWith("\n"))
+              writeln();
+        writeln("==============================");
+        remove(output_file_temp);
+    }
+
+    rename(output_file_temp, output_file);
+    // TODO: should we remove all the test artifacts if the test passes? rmdirRecurse(testOutDir)?
+    return exitCode;
+}
+
+/// return the dshell_prebuilt object file
+string getDShellPrebuilt(const ref EnvData envData)
+{
+    return buildPath(envData.results_dir, "dshell_prebuilt" ~ envData.obj);
+}
+
+/// build the dshell prebuilt
+int buildDShellPrebuilt(const ref EnvData envData)
```

Comment:
When I compiled `dshell_prebuilt` with `run.d` (which I started with) it doesn't link to the code compiled with `d_do_test`.  Since `d_do_test` is the one compiling the dshell tests, I realized it would be simpler to have `d_do_test` also compile `dshell_prebuilt` so we can ensure it compiles it in the same environment and with the same flags as the `dshell` tests themselves.

I could try again to get this to work with `run.d`, but I'll probably need to copy some code from `d_do_test`.  Also, the makefile will need to copy over whatever logic I put into `run.d`.  Right now they both just call `d_do_test -dshell_prebuilt` and all the logic is common and stays in `d_do_shell`.

Reply:
Same thing as the my previous comment.  This is no longer relevant as I got `run.d` and `Makefile` to build it directly now.